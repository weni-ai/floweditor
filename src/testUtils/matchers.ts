import { commaListsOr } from 'common-tags';
import { Exit } from 'flowTypes';
import { RenderNode } from 'store/flowContext';

const snapshot = require('jest-snapshot');

const matchers: jest.ExpectExtendMap = {};

interface MatchResult {
  message: () => string;
  pass: boolean;
}

function toBeUnique<T>(this: jest.MatcherUtils, received: any[]): MatchResult {
  const seen: { [value: string]: boolean } = {};

  for (const item of received) {
    const stringified = JSON.stringify(item, null, 1);
    if (seen[stringified]) {
      return {
        message: () => `Duplicate item in array:\n${stringified}`,
        pass: false,
      };
    } else {
      seen[stringified] = true;
    }
  }

  return {
    message: () => 'Array is unique',
    pass: true,
  };
}

function toMatchCallSnapshot<T>(
  this: jest.MatcherUtils,
  received: any,
  snapshotName?: string,
): MatchResult {
  return snapshot.toMatchSnapshot.call(
    this,
    received.mock.calls[0],
    snapshotName,
  );
}

function toHaveInboundConnections<T>(
  this: jest.MatcherUtils,
  received: RenderNode,
): MatchResult {
  if (Object.keys(received.inboundConnections).length > 1) {
    return {
      message: () => `Node ${received.node.uuid} has an inbound connection`,
      pass: true,
    };
  }

  return {
    message: () => `Node ${received.node.uuid} has no inbound connections`,
    pass: false,
  };
}

function toHaveInboundFrom<T>(
  this: jest.MatcherUtils,
  received: RenderNode,
  expected: Exit,
): MatchResult {
  for (const exitUUID of Object.keys(received.inboundConnections)) {
    if (exitUUID === expected.uuid) {
      return {
        message: () =>
          `Node ${received.node.uuid} incorrectly has inbound from ${expected.uuid}`,
        pass: true,
      };
    }
  }

  return {
    message: () =>
      `Node ${received.node.uuid} doesn't have inbound connection from ${expected.uuid}`,
    pass: false,
  };
}

function toPointTo<T>(
  this: jest.MatcherUtils,
  received: Exit,
  expected: RenderNode,
): MatchResult {
  if (received.destination_uuid === expected.node.uuid) {
    return {
      message: () =>
        `${received.uuid} incorrectly points to ${expected.node.uuid}`,
      pass: true,
    };
  }

  return {
    message: () =>
      `Expected ${received.uuid} to point to ${expected.node.uuid}`,
    pass: false,
  };
}

function toHaveExitWithDestination<T>(
  this: jest.MatcherUtils,
  received: RenderNode,
): MatchResult {
  for (const exit of received.node.exits) {
    if (exit.destination_uuid !== null) {
      return {
        message: () =>
          `Exit ${exit.uuid} in node ${received.node.uuid} has a destination`,
        pass: true,
      };
    }
  }

  return {
    message: () => `All exits for ${received.node.uuid} are disconnected`,
    pass: false,
  };
}

function toHaveExitThatPointsTo<T>(
  this: jest.MatcherUtils,
  received: RenderNode,
  expected: RenderNode,
): MatchResult {
  for (const exit of received.node.exits) {
    if (exit.destination_uuid === expected.node.uuid) {
      return {
        message: () =>
          `${exit.uuid} incorrectly has exit that points to ${expected.node.uuid}`,
        pass: true,
      };
    }
  }

  return {
    message: () =>
      `Expected ${received.node.uuid} to have an exit that points to ${expected.node.uuid}`,
    pass: false,
  };
}

function toHavePayload<T>(
  this: jest.MatcherUtils,
  store: any,
  actionType: string,
  expectedPayload: any,
): MatchResult {
  const payload = JSON.stringify(expectedPayload);
  for (const actionTaken of store.getActions()) {
    if (actionTaken.type === actionType) {
      if (JSON.stringify(actionTaken.payload) === payload) {
        return {
          message: () =>
            `Result contained action type ${actionType} with payload ${payload}`,
          pass: true,
        };
      }
    }
  }

  return {
    message: () =>
      `Could not find action type ${actionType} with payload ${payload}`,
    pass: false,
  };
}

function toHaveReduxActions<T>(
  this: jest.MatcherUtils,
  store: any,
  actionTypes: string[],
): MatchResult {
  const missedTypes: any[] = [];
  for (const actionTaken of store.getActions()) {
    for (const actionType of actionTypes) {
      if (actionTaken.type === actionType) {
        return {
          message: () => `Result contained action type ${actionType}`,
          pass: true,
        };
      } else {
        missedTypes.push(actionType);
      }
    }
  }

  return {
    message: () =>
      commaListsOr`Could not find action type(s): ${[...missedTypes]}`,
    pass: false,
  };
}

expect.extend({
  toPointTo,
  toHaveExitThatPointsTo,
  toHaveInboundFrom,
  toHaveExitWithDestination,
  toHaveInboundConnections,
  toHavePayload,
  toHaveReduxActions,
  toMatchCallSnapshot,
  toBeUnique,
});
