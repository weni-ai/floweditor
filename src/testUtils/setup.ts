/* eslint-disable no-redeclare */
import { Console } from 'console';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Exit } from 'flowTypes';
import { RenderNode } from 'store/flowContext';

// Declare custom matcher types
declare global {
  namespace jest {
    interface Matchers<R> {
      toPointTo(renderNode: RenderNode): R;
      toHaveExitThatPointsTo(renderNode: RenderNode): R;
      toHaveInboundFrom(exit: Exit): R;
      toHaveExitWithDestination(): R;
      toHaveInboundConnections(): R;
      toHavePayload(action: string, payload: any): R;
      toHaveReduxActions(actions: string[]): R;
      toMatchCallSnapshot(snapshotName?: string): R;
    }
  }
}

// Ensure console logs are visible while running tests https://github.com/facebook/jest/issues/3853
global.console = new Console(process.stderr, process.stderr);

// Configure Enzyme adapter
configure({ adapter: new Adapter() });

// RAF shim
// tslint:disable-next-line:ban-types
(global as any).requestAnimationFrame = (callback: Function) => {
  setTimeout(callback, 0);
};
