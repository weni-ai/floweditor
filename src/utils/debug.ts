import { FlowStoreProps } from 'components/flow/Flow';
import { DebugState } from 'store/editor';

import mutate from 'immutability-helper';

/* istanbul ignore next -- @preserve */
export default class Debug {
  private props: FlowStoreProps;
  private state: DebugState;

  constructor(props: FlowStoreProps, initial: DebugState) {
    this.props = props;
    this.state = initial || { showUUIDs: false };
  }

  public showUUIDs(): DebugState {
    const updated = mutate(this.state, { $merge: { showUUIDs: true } });
    this.props.mergeEditorState({ debug: updated });
    return updated;
  }
}
