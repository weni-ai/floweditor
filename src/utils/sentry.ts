import * as Sentry from '@sentry/react';

const initSentry = (dsn: string) => {
  Sentry.init({
    dsn,
    integrations: [
      Sentry.browserTracingIntegration(),
      Sentry.httpClientIntegration({
        failedRequestStatusCodes: [[400, 499], [500, 599]],
      }),
      Sentry.replayIntegration(),
      Sentry.replayCanvasIntegration(),
    ],
    tracesSampleRate: 1.0,
    replaysSessionSampleRate: 0.1,
    replaysOnErrorSampleRate: 1.0,
  });
};

export default initSentry;
