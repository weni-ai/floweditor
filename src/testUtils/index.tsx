// TODO: Remove use of Function
// tslint:disable:ban-types
import { ConfigProviderContext, fakePropType } from 'config/ConfigProvider';
import { mount, ReactWrapper, shallow, ShallowWrapper } from 'enzyme';
import { FlowDefinition, FlowEditorConfig } from 'flowTypes';
import mutate, { Query } from 'immutability-helper';
import * as React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import createStore from 'store/createStore';
import { RenderNodeMap } from 'store/flowContext';
import { getFlowComponents } from 'store/helpers';
import AppState, { initialState } from 'store/state';
import config from 'test/config';
import * as matchers from 'testUtils/matchers';
import { merge, set } from 'utils';

// we need to use require syntax to bust implicit any
const boring: FlowDefinition = require('test/flows/boring.json');

// force our matchers to be read in
const match = matchers;

export interface Resp {
  assets: { [key: string]: any }[];
}

export interface QueryString {
  [key: string]: string;
}

export const contextTypes: { [key: string]: Function } = {
  config: fakePropType,
  store: fakePropType,
  assetService: fakePropType,
};

export const baseState: AppState = mutate(initialState, {
  flowContext: merge({
    definition: require('test/flows/colors.json') as FlowDefinition,
  }),
  editorState: merge({
    language: { id: 'eng', name: 'English' },
  }),
});

const flowEditorConfig: FlowEditorConfig = config;

export const configProviderContext: ConfigProviderContext = {
  config: flowEditorConfig,
};

export const setMock = (
  implementation?: (...args: any[]) => any,
): Query<jest.Mock> => set(jest.fn(implementation));

/**
 * Compose setup method for component tests
 */
export const composeSetup = <P extends {}>(
  Component: React.ComponentClass | React.SFC,
  baseProps: P = {} as any,
  baseDuxState: AppState | Partial<AppState> = baseState,
  baseContext: ConfigProviderContext = configProviderContext,
) => (
  shallowRender: boolean = true,
  propOverrides: Query<P | Partial<P>> = {},
  duxStateOverrides: Query<AppState | Partial<AppState>> = {},
  contextOverrides: Query<
    ConfigProviderContext | Partial<ConfigProviderContext>
  > = {},
  childContextTypeOverrides: { [key: string]: Function } = {},
  children: (JSX.Element | React.ComponentClass | React.SFC)[] = [],
) => {
  const props = mutate(baseProps, propOverrides);
  const store = createStore(mutate(
    baseDuxState,
    duxStateOverrides,
  ) as AppState);
  let context = mutate(baseContext, merge({ store }));

  if (Object.keys(duxStateOverrides).length > 0) {
    context = mutate(context, duxStateOverrides);
  }

  if (Object.keys(contextOverrides).length > 0) {
    context = mutate(context, contextOverrides as ConfigProviderContext);
  }

  const childContextTypes: { [contextProp: string]: Function } = mutate(
    contextTypes,
    childContextTypeOverrides,
  );
  // tslint:disable-next-line:ban-types
  const wrapper = (shallowRender ? (shallow as Function) : (mount as Function))(
    <Component {...props}>{children}</Component>,
    {
      context,
      childContextTypes,
    },
  );

  return {
    wrapper,
    props,
    context,
    instance: wrapper.instance(),
  };
};

export const composeSpy = (obj: Record<string, any> | React.ComponentClass) => (
  instanceMethod: string,
) =>
  jest.spyOn(
    (obj as React.ComponentClass).prototype || obj,
    instanceMethod as any,
  );

/**
 * Wait for promises in queue to resolve
 */
export const flushPromises = () =>
  new Promise((resolve: any) => setImmediate(resolve));

/**
 * Restore spy mocks (distinct from mocks created w/ jest.fn()).
 * Use to declaratively restore multiple spy mocks,
 * otherwise just call .mockRestore() on lone spy.
 */
export const restoreSpies = (...spies: jest.SpyInstance<any>[]) =>
  spies.forEach((spy: any) => spy.mockRestore());

// To-do: type this method's output, can pass it prop generic ingested by getComponentTestUtils
/**
 * NOTE: borrowed from EventBrite: https://github.com/eventbrite/javascript/blob/master/react/testing.md#finding-nodes
 * Finds all instances of components in the rendered `componentWrapper` that are DOM components
 * with the `data-spec` attribute matching `name`.
 * @param {ReactWrapper|ShallowWrapper} componentWrapper - Rendered componentWrapper (result of mount, shallow, or render)
 * @param  {string} snacName - Name of `data-spec` attribute value to find
 * @param {string|Function} (Optional) typeFilter - (Optional) Expected type of the wrappers (defaults to all HTML tags)
 * @returns {ReactWrapper|ReactWrapper[]|ShallowWrapper|ShallowWrapper[]} Matching DOM components
 */
export const getSpecWrapper = (
  componentWrapper: ReactWrapper<{}, {}> | ShallowWrapper<{}, {}>,
  specName: string,
  attributeName: string = 'data-spec',
): any => componentWrapper.find(`[${attributeName}="${specName}"]`);

export const composeDuxState = (
  query: Query<AppState | Partial<AppState>> = {},
  duxState = baseState,
) => mutate(duxState, query);

export const composeComponentTestUtils = <P extends {}>(
  Component: React.ComponentClass | React.SFC | any,
  baseProps: P = {} as any,
  baseDuxState: AppState | Partial<AppState> = baseState,
  baseContext: ConfigProviderContext = configProviderContext,
) => ({
  setup: composeSetup<P>(Component, baseProps, baseDuxState, baseContext),
  spyOn: composeSpy(Component),
});

export const prepMockDuxState = (): {
  testNodes: RenderNodeMap;
  mockDuxState: AppState;
} => {
  const testNodes = getFlowComponents(boring).renderNodeMap;

  return {
    testNodes,
    mockDuxState: {
      ...initialState,
      flowContext: {
        ...initialState.flowContext,
        definition: boring,
        nodes: testNodes,
      },
    },
  };
};

export const createMockStore: Function = configureStore([thunk]);

export const mock = <T extends {}, K extends keyof T>(
  object: T,
  property: K,
  value: T[K],
) => {
  Object.defineProperty(object, property, { get: () => value });
};
