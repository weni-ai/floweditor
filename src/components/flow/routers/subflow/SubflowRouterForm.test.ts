import SubflowRouterForm from 'components/flow/routers/subflow/SubflowRouterForm';
import { composeComponentTestUtils, mock, setMock } from 'testUtils';
import {
  ColorFlowAsset,
  createStartFlowAction,
  createSubflowNode,
  getRouterFormProps,
} from 'testUtils/assetCreators';
import * as utils from 'utils';

const routerNode = createSubflowNode(createStartFlowAction());
const { setup } = composeComponentTestUtils(
  SubflowRouterForm,
  getRouterFormProps(routerNode),
);
mock(utils, 'createUUID', utils.seededUUIDs());

describe(SubflowRouterForm.name, () => {
  describe('render', () => {
    it('should render', () => {
      const { wrapper } = setup();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('updates', () => {
    it('should update and save', () => {
      const { instance, props } = setup(true, { updateRouter: setMock() });
      instance.handleFlowChanged([ColorFlowAsset]);
      expect(instance.state).toMatchSnapshot();
      instance.getButtons().primary.onClick();
      expect(props.updateRouter).toMatchCallSnapshot();
    });

    it('should cancel changes', () => {
      const { instance, props } = setup(true, { updateRouter: setMock() });
      instance.handleFlowChanged([ColorFlowAsset]);
      instance.getButtons().secondary.onClick();
      expect(props.updateRouter).not.toBeCalled();
    });

    it('converts from other node types', () => {
      const { instance, props } = setup(true, {
        updateRouter: setMock(),
        nodeSettings: {
          originalNode: { ui: { $merge: { type: null } } },
        },
      });

      instance.handleFlowChanged([ColorFlowAsset]);
      instance.getButtons().primary.onClick();
      expect(props.updateRouter).toMatchCallSnapshot();
    });

    it('creates its own action uuid if necessary', () => {
      const { instance, props } = setup(true, {
        updateRouter: setMock(),
        nodeSettings: {
          originalNode: {
            node: { $merge: { actions: [] } },
            ui: { $merge: { type: null } },
          },
        },
      });

      instance.handleFlowChanged([ColorFlowAsset]);
      instance.getButtons().primary.onClick();
      expect(props.updateRouter).toMatchCallSnapshot();
    });

    it('validates before saving', () => {
      const { instance, props } = setup(true, { updateRouter: setMock() });
      instance.handleFlowChanged([]);
      instance.getButtons().primary.onClick();
      expect(props.updateRouter).not.toBeCalled();
    });
  });
});
