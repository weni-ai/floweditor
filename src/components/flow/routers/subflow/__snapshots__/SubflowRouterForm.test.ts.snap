// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SubflowRouterForm > render > should render 1`] = `
<Dialog
  buttons={
    {
      "primary": {
        "name": "Confirm",
        "onClick": [Function],
      },
      "secondary": {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  className="_dialog_40da5e"
  headerClass="enter_flow"
  tabs={[]}
  title="Enter another flow"
>
  <TypeList
    __className=""
    initialType={
      {
        "aliases": [
          "split_by_subflow",
        ],
        "component": [Function],
        "description": "Enter another flow",
        "form": [Function],
        "localization": [Function],
        "localizeableKeys": [
          "exits",
        ],
        "name": "Enter another flow",
        "type": "enter_flow",
      }
    }
    nodeSettings={
      {
        "originalAction": null,
        "originalNode": {
          "inboundConnections": {},
          "node": {
            "actions": [
              {
                "flow": {
                  "name": "Colors",
                  "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
                },
                "type": "enter_flow",
                "uuid": "d4a3a01c-1dee-4324-b107-4ac7a21d836f",
              },
            ],
            "exits": [
              {
                "destination_uuid": null,
                "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
              },
              {
                "destination_uuid": null,
                "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
              },
            ],
            "router": {
              "cases": [
                {
                  "arguments": [
                    "C",
                  ],
                  "category_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
                  "type": "has_run_status",
                  "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
                },
                {
                  "arguments": [
                    "E",
                  ],
                  "category_uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
                  "type": "has_run_status",
                  "uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
                },
              ],
              "categories": [
                {
                  "exit_uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
                  "name": "Complete",
                  "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
                },
                {
                  "exit_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
                  "name": "Expired",
                  "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
                },
              ],
              "default_category_uuid": null,
              "operand": "@child",
              "type": "switch",
              "wait": null,
            },
            "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
          },
          "ui": {
            "position": {
              "left": 0,
              "top": 0,
            },
            "type": "split_by_subflow",
          },
        },
      }
    }
    onChange={[MockFunction spy]}
  />
  <AssetSelector
    assets={
      {
        "items": {},
        "type": "flow",
      }
    }
    entry={
      {
        "value": {
          "name": "Colors",
          "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
        },
      }
    }
    name="Flow to start"
    onChange={[Function]}
    placeholder="Select the flow to start"
    searchable={true}
    shouldExclude={[Function]}
    showLabel={true}
  />
</Dialog>
`;

exports[`SubflowRouterForm > updates > converts from other node types 1`] = `
[MockFunction spy] {
  "calls": [
    [
      {
        "inboundConnections": {},
        "node": {
          "actions": [
            {
              "flow": {
                "name": "Favorite Color",
                "uuid": "9a93ede6-078f-44c9-ad0a-133793be5d56",
              },
              "type": "enter_flow",
              "uuid": "d4a3a01c-1dee-4324-b107-4ac7a21d836f",
            },
          ],
          "exits": [
            {
              "destination_uuid": null,
              "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
            },
            {
              "destination_uuid": null,
              "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
            },
          ],
          "router": {
            "cases": [
              {
                "arguments": [
                  "completed",
                ],
                "category_uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
                "type": "has_only_text",
                "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
              },
              {
                "arguments": [
                  "expired",
                ],
                "category_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
                "type": "has_only_text",
                "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
              },
            ],
            "categories": [
              {
                "exit_uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
                "name": "Complete",
                "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
              },
              {
                "exit_uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
                "name": "Expired",
                "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
              },
            ],
            "default_category_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
            "operand": "@child.status",
            "type": "switch",
          },
          "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
        },
        "ui": {
          "config": {},
          "position": null,
          "type": "split_by_subflow",
        },
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`SubflowRouterForm > updates > creates its own action uuid if necessary 1`] = `
[MockFunction spy] {
  "calls": [
    [
      {
        "inboundConnections": {},
        "node": {
          "actions": [
            {
              "flow": {
                "name": "Favorite Color",
                "uuid": "9a93ede6-078f-44c9-ad0a-133793be5d56",
              },
              "type": "enter_flow",
              "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
            },
          ],
          "exits": [
            {
              "destination_uuid": null,
              "uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
            },
            {
              "destination_uuid": null,
              "uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
            },
          ],
          "router": {
            "cases": [
              {
                "arguments": [
                  "completed",
                ],
                "category_uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
                "type": "has_only_text",
                "uuid": "e0bc3776-792f-49d2-b4be-48be2cf8159a",
              },
              {
                "arguments": [
                  "expired",
                ],
                "category_uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
                "type": "has_only_text",
                "uuid": "6c22884b-0e35-4206-982e-18320691eda9",
              },
            ],
            "categories": [
              {
                "exit_uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
                "name": "Complete",
                "uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
              },
              {
                "exit_uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
                "name": "Expired",
                "uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
              },
            ],
            "default_category_uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
            "operand": "@child.status",
            "type": "switch",
          },
          "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
        },
        "ui": {
          "config": {},
          "position": null,
          "type": "split_by_subflow",
        },
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`SubflowRouterForm > updates > should update and save 1`] = `
{
  "flow": {
    "validationFailures": [],
    "value": {
      "name": "Favorite Color",
      "parent_refs": [
        "colors",
      ],
      "uuid": "9a93ede6-078f-44c9-ad0a-133793be5d56",
    },
  },
  "params": {
    "colors": {
      "value": "",
    },
  },
  "valid": true,
}
`;

exports[`SubflowRouterForm > updates > should update and save 2`] = `
[MockFunction spy] {
  "calls": [
    [
      {
        "inboundConnections": {},
        "node": {
          "actions": [
            {
              "flow": {
                "name": "Favorite Color",
                "uuid": "9a93ede6-078f-44c9-ad0a-133793be5d56",
              },
              "type": "enter_flow",
              "uuid": "d4a3a01c-1dee-4324-b107-4ac7a21d836f",
            },
          ],
          "exits": [
            {
              "destination_uuid": null,
              "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
            },
            {
              "destination_uuid": null,
              "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
            },
          ],
          "router": {
            "cases": [
              {
                "arguments": [
                  "C",
                ],
                "category_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
                "type": "has_run_status",
                "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
              },
              {
                "arguments": [
                  "E",
                ],
                "category_uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
                "type": "has_run_status",
                "uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
              },
            ],
            "categories": [
              {
                "exit_uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
                "name": "Complete",
                "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
              },
              {
                "exit_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
                "name": "Expired",
                "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
              },
            ],
            "default_category_uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
            "operand": "@child.status",
            "type": "switch",
          },
          "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
        },
        "ui": {
          "config": {},
          "position": null,
          "type": "split_by_subflow",
        },
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
